File: scripts/producer/tests/live_grade_v2_supervisor.ts
/** TEST-only blocked supervisor pipe. No claim or inner worker exists at spawn. */
import { spawn, type ChildProcessWithoutNullStreams } from "node:child_process";
import path from "node:path";
import { watchGradeProcess, translatedDeadline, type GradeInvocation } from "../../../src/lib/server/grade-observation-process";
import { checkTime } from "../../../src/lib/server/grade-observation-store";
import { exactIdentity, groupState } from "./live_grade_v2_lifecycle";
import type { V2LiveReturn } from "./live_grade_v2_finalization";

export interface V2Supervisor {
  child: ChildProcessWithoutNullStreams; identity: ReturnType<typeof exactIdentity>; deadlineNs: string;
  closed: { code: number | null; signal: string | null; absent: boolean }; close: Promise<void>;
}
async function waitClose(close: Promise<void>, ms: number): Promise<void> {
  let timer: ReturnType<typeof setTimeout> | undefined;
  try { await Promise.race([close, new Promise<void>(resolve => { timer = setTimeout(resolve, ms); })]); }
  finally { clearTimeout(timer); }
}
function ready(child: ChildProcessWithoutNullStreams, deadline: bigint): Promise<string> {
  return new Promise((resolve, reject) => {
    let raw = "";
    const cleanup = () => { clearTimeout(timer); child.stdout.off("data", data); child.off("error", failure); child.off("close", failure); };
    const failure = () => { cleanup(); reject(new Error("Blocked supervisor startup failed")); };
    const data = (part: Buffer) => {
      raw += part.toString("utf8");
      if (raw.length > 4096) { failure(); return; }
      if (!raw.includes("\n")) return;
      try {
        const value = JSON.parse(raw);
        if (Object.keys(value).sort().join() !== "readyMonotonicNs,supervisorReady" || value.supervisorReady !== child.pid)
          throw new Error("Invalid supervisor ready");
        const translated = translatedDeadline(value.readyMonotonicNs, deadline);
        cleanup(); resolve(translated);
      } catch { failure(); }
    };
    const timer = setTimeout(failure, Math.max(1, Math.min(5000, Number(deadline - process.hrtime.bigint()) / 1e6)));
    child.stdout.on("data", data); child.once("error", failure); child.once("close", failure);
  });
}
export async function startSupervisor(input: { root: string; python: string; deadline: bigint }): Promise<V2Supervisor> {
  checkTime(input.deadline);
  const waitMs = Math.floor(Number(input.deadline - process.hrtime.bigint()) / 1e6);
  const script = path.join(input.root, "scripts/producer/tests/live_grade_v2_supervisor.py");
  const child = spawn(input.python, ["-I", "-S", "-B", script, String(process.pid), String(waitMs)],
    { cwd: input.root, env: { ...process.env }, detached: true, stdio: ["pipe", "pipe", "pipe"] });
  const closed = { code: null as number | null, signal: null as string | null, absent: false };
  const close = new Promise<void>(resolve => child.once("close", (code, signal) => {
    Object.assign(closed, { code, signal, absent: !!child.pid && groupState(child.pid) === "absent" }); resolve();
  }));
  try {
    const deadlineNs = await ready(child, input.deadline);
    const identity = exactIdentity(child.pid!); checkTime(input.deadline);
    return { child, identity, deadlineNs, closed, close };
  } catch (error) {
    // No permit was sent: EOF cannot fork. Wait for the bounded blocked child.
    child.stdin.end(); await waitClose(close, 1500);
    if (child.pid && groupState(child.pid) !== "absent") throw new AggregateError([error], "Unstarted supervisor absence unproved");
    throw error;
  }
}
function same(value: V2Supervisor): boolean {
  try { return JSON.stringify(exactIdentity(value.identity.pid)) === JSON.stringify(value.identity); }
  catch { return false; }
}

/** Monitor is attached before the one permitted activation line is sent. */
export async function invokeSupervisor(value: V2Supervisor, input: GradeInvocation,
  permit: { directory: string; lifecycleSha256: string; claimPath: string; claimSha256: string }): Promise<V2LiveReturn> {
  const logs = { stdout: Buffer.alloc(0), stderr: Buffer.alloc(0) }, child = value.child;
  for (const channel of ["stdout", "stderr"] as const) child[channel].on("data", (part: Buffer) => {
    logs[channel] = Buffer.concat([logs[channel], part.subarray(0, 128 * 1024 - logs[channel].length)]);
  });
  const pending = watchGradeProcess(input, child, { identity: () => same(value), now: process.hrtime.bigint,
    signal: signal => { if (same(value)) process.kill(signal === "SIGKILL" ? -value.identity.pid : value.identity.pid, signal); },
    groupAlive: () => groupState(value.identity.pid) !== "absent", timer: setTimeout, clear: clearTimeout });
  child.stdin.write(JSON.stringify(permit) + "\n");
  const held = await pending;
  await waitClose(value.close, 2000);
  return { held, closed: { ...value.closed, absent: groupState(value.identity.pid) === "absent" },
    pid: value.identity.pid, stdout: logs.stdout.toString("utf8"), stderr: logs.stderr.toString("utf8") };
}

/** Setup errors close the unactivated pipe; no container can have been launched. */
export async function closeBlockedSupervisor(value: V2Supervisor): Promise<void> {
  value.child.stdin.end(); await waitClose(value.close, 1500);
  if (groupState(value.identity.pid) !== "absent") throw new Error("Blocked supervisor absence unproved");
}

File: scripts/producer/tests/live_grade_v2_supervisor.py
"""TEST-only pre-claim blocked session; fixed worker fork only after exact permit.

The outer owner independently observes group absence. This supervisor never
claims that its own exit proves absence of descendants or Docker resources.
"""
from __future__ import annotations

import json
import os
import select
import signal
import subprocess
import sys
import time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
from cut_preview_io import bound_json, file_hash, real_directory


def _permit_object(raw: bytes) -> dict:
    """Reject nonobject activation without a nested wait-loop branch."""
    value = json.loads(raw)
    if type(value) is not dict:
        raise RuntimeError("Fork permit must be an object")
    return value


def read_permit(parent: int, deadline: float) -> dict:
    """Read one bounded line without prefetching the subsequent worker handshake."""
    raw = bytearray()
    while time.monotonic() < deadline and os.getppid() == parent:
        if not select.select([0], [], [], min(.1, max(0, deadline - time.monotonic())))[0]:
            continue
        part = os.read(0, 1)
        if not part:
            raise RuntimeError("Owner pipe closed before fork permit")
        raw.extend(part)
        if part == b"\n":
            return _permit_object(raw)
        if len(raw) > 4096:
            raise RuntimeError("Fork permit exceeds bound")
    raise RuntimeError("Owner absent or original fork deadline expired")


def validate_permit(row: dict, parent: int) -> tuple[list[str], float]:
    """Cross-bind immutable lifecycle, exact active claim and input before fork."""
    if set(row) != {"directory", "lifecycleSha256", "claimPath", "claimSha256"}:
        raise RuntimeError("Malformed exact fork permit")
    directory = Path(row["directory"])
    real_directory(directory)
    life = bound_json(directory / "TEST-lifecycle.json", row["lifecycleSha256"])
    claim = bound_json(Path(row["claimPath"]), row["claimSha256"])
    expected = {"jobId": directory.name, "dir": life["producerDir"],
                "requestDigest": life["inputSha256"], "lifecycleSha256": row["lifecycleSha256"]}
    if claim != expected or sys.executable != life["python"] \
            or row["claimPath"] != str(Path(life["resource"]) / "active.json") \
            or life["directory"] != str(directory) \
            or life["supervisor"]["pid"] != os.getpid() or life["owner"]["pid"] != parent:
        raise RuntimeError("Fork permit differs from original owner claim")
    for pin in life["pins"]:
        if file_hash(Path(pin["path"]), 128 * 1024 * 1024) != pin["sha256"]:
            raise RuntimeError("Held supervisor/worker code changed before fork")
    value = bound_json(directory / "input.json", life["inputSha256"])
    if value["ownerPid"] != os.getpid() or value["jobId"] != directory.name:
        raise RuntimeError("Worker input does not name this supervisor")
    script = Path(__file__).with_name("live_grade_v2_launch.py")
    command = [sys.executable, "-I", "-S", "-B", str(script), str(directory), row["lifecycleSha256"]]
    return command, int(life["supervisorDeadlineNs"]) / 1e9


def supervise(command: list[str], parent: int, deadline: float, cleanup_seconds: float = 100) -> int:
    """Keep leaf children in this group; cancel on owner death without new work credit."""
    if os.getppid() != parent or time.monotonic() >= deadline or not 0 < cleanup_seconds <= 100:
        raise RuntimeError("Original owner or fork deadline unavailable")
    child = subprocess.Popen(command, start_new_session=False)
    cancelled = False
    def cancel(_sig: int, _frame: object) -> None:
        """Forward work cancellation only to the actual unreaped child."""
        nonlocal cancelled
        if child.poll() is None:
            child.send_signal(signal.SIGUSR1)
        cancelled = True
    signal.signal(signal.SIGUSR1, cancel)
    signal.signal(signal.SIGTERM, cancel)
    while child.poll() is None:
        if not cancelled and (os.getppid() != parent or time.monotonic() >= deadline):
            cancel(signal.SIGUSR1, None)
        if time.monotonic() >= deadline + cleanup_seconds:
            child.kill()
            return child.wait(timeout=2)
        time.sleep(.02)
    return child.returncode


def main() -> None:
    """Only a POSIX session leader with a live pipe owner may await activation."""
    if len(sys.argv) != 3 or not sys.argv[1].isdigit() or not sys.argv[2].isdigit():
        raise RuntimeError("Internal TEST supervisor requires parent and bounded wait")
    parent, wait_ms = int(sys.argv[1]), int(sys.argv[2])
    if parent != os.getppid() or os.getpid() != os.getpgrp() or os.getpid() != os.getsid(0) \
            or not 1 <= wait_ms <= 120000:
        raise RuntimeError("Invalid owned supervisor group or deadline")
    sample = time.monotonic_ns()
    deadline = sample / 1e9 + wait_ms / 1000
    print(json.dumps({"supervisorReady": os.getpid(), "readyMonotonicNs": str(sample)}), flush=True)
    permit = read_permit(parent, deadline)
    command, held_deadline = validate_permit(permit, parent)
    if not time.monotonic() < held_deadline <= deadline:
        raise RuntimeError("Original translated supervisor deadline unavailable")
    raise SystemExit(supervise(command, parent, held_deadline))


if __name__ == "__main__":
    main()
